ARROW-4104: [Java] fix a race condition in AllocationManager

pravindra opened the pull request.

pravindra commented:
@siddharthteotia @jacques-n can you please review ?

1. I changed the default allocation size to 3968 (from 4096) since allocating 4096 elements causes a wastage of at ~16K.
2. Lot of tests assume an initial allocation of 4096 or 4095 - I've modified the tests to explicitly set the initial allocation size.
3. fixed a race condition (bug in my earlier change)

jacques-n commented on java/vector/src/main/java/org/apache/arrow/vector/BaseValueVector.java:
```diff
@@ -98,5 +104,40 @@ protected ArrowBuf releaseBuffer(ArrowBuf buffer) {
+  /*
+   * Allocate a single buffer and slice into multiple.
+   */
+  protected static ArrowBuf[] allocAndSlice(BufferAllocator allocator, int[] lengths) {
```
This seems more generic than necessary. Why not make it size aware and do the allocation based on a provided value count? Also, if int[] is always length two in usage, let's make it focus on that case. For example, we shouldn't be allocating the variable data chunk in the same vector since it will have a different lifecycle.

pravindra replied:
I want to use this for both fixed-len (data+validity) and var-len vectors (offsets+validity). 

- For the fixed-len vectors, the data buffers and validity buffers are sized for the same number-of-elements. 
- For the var-len vectors, the offset buffer needs to be sized for one extra element than the bitmap.

jacques-n commented on java/vector/src/main/java/org/apache/arrow/vector/BaseVariableWidthVector.java:
```diff
@@ -522,15 +530,44 @@ public void reallocDataBuffer() {
   public void reallocValidityAndOffsetBuffers() {
-    offsetBuffer = reallocBufferHelper(offsetBuffer, true);
-    validityBuffer = reallocBufferHelper(validityBuffer, false);
+    int offsetBaseSize = Integer.max(offsetBuffer.capacity(), offsetAllocationSizeInBytes);
```
I don't understand why this can't be generic just like the the original alloc with just valuecount *2

pravindra replied:
thinking further, I think this can be simplified significantly if I make both the initial allocation and the reAlloc target a multiple-of-2 allocation. 

I'll need to fix a lot of tests (the allocation may be larger than what was asked for), but it's probably worth the simplicity.

jacques-n commented on java/vector/src/main/java/org/apache/arrow/vector/BaseVariableWidthVector.java:
```diff
@@ -522,15 +530,44 @@ public void reallocDataBuffer() {
-    long baseSize = (offsetBuffer ? offsetAllocationSizeInBytes
-            : validityAllocationSizeInBytes);
+  private void reallocValidityBuffer() {
+    final int currentBufferCapacity = validityBuffer.capacity();
```
Why do we have this method at all? Shouldn't this be shared with the value buffer?

pravindra replied:
This is used by an api called setNull(int index) -  it just sets the validity bit to false without actually adding a data element. If this optimization isn't useful, I can switch to expand both validity/offset arrays (I preserved the old behavior).

Also, used by another api setIndexDefined(int index). I think this one isn't meaningful unless we expand both the validity and the offset arrays.

jacques-n replied:
Yes, we should always only resize together. If you resized the validity buffer, you'll need to resize the data buffer ultimately anyway (when you call setValueCount). We shouldn't be ever doing a partial resize.

pravindra commented:
@jacques-n - I've addressed the review comments. There is one change one behavior - both allocateNew and reAlloc can allocate more than the requested size, to ensure that the allocated buffer size is a power-of-2. Most of the test related changes are related to this.

can you please review now ?

pravindra commented on java/vector/src/main/java/org/apache/arrow/vector/BaseFixedWidthVector.java:
```diff
@@ -159,12 +151,11 @@ public ArrowBuf getOffsetBuffer() {
     }
-    valueAllocationSizeInBytes = (int) size;
-    validityAllocationSizeInBytes = getValidityBufferSizeFromCount(valueCount);
+    initialValueAllocation = valueCount;
```
no, the value count isn't rounded up. The buffer size is rounded up to a power-of-2 when the allocation happens, and the buffer is then used to it's full capacity.

pravindra commented on java/vector/src/main/java/org/apache/arrow/vector/BaseFixedWidthVector.java:
```diff
@@ -267,18 +258,15 @@ public void allocateNew() {
-
-    if (align(curAllocationSizeValue) + curAllocationSizeValidity > MAX_ALLOCATION_SIZE) {
+    if (computeCombinedBufferSize(initialValueAllocation, typeWidth) > MAX_ALLOCATION_SIZE) {
       throw new OversizedAllocationException("Requested amount of memory exceeds limit");
```
sorry, are you asking more detail in the error message, like

B bytes of memory required for capacity of C values, which exceeds the max limit of L bytes

pravindra commented on java/vector/src/main/java/org/apache/arrow/vector/BaseValueVector.java:
```diff
@@ -98,5 +105,67 @@ protected ArrowBuf releaseBuffer(ArrowBuf buffer) {
+    return BaseAllocator.nextPowerOfTwo(bufferSize);
+  }
+
+  protected ArrowBuf[] allocDataAndValidityBufs(int valueCount, int typeWidth) {
```
will change name

pravindra commented on java/vector/src/main/java/org/apache/arrow/vector/BaseValueVector.java:
```diff
@@ -98,5 +105,67 @@ protected ArrowBuf releaseBuffer(ArrowBuf buffer) {
+    return ((size + 7) / 8) * 8;
+  }
+
+  protected long computeCombinedBufferSize(int valueCount, int typeWidth) {
```
I'll add the pre-checks.

> Also, shouldn't we be returning the calculated valuecount? e.g. if you size for 50 but it turns > out we can fit 60 in the same memory, shouldn't we be using that as the available size/value > count (for initial alloc purposes, etc)?

That's done as part of allocDataAndValidityBufs (both for initial alloc and realloc) - in this case, it will return buffers with capacity of 60. And, the vector implementations will utilize the total capacity of 60.

The api setInitialCapacity() sets the variables based on whatever the user requests. Later, during alloc/realloc, the buffers allocated can have a capacity greater than this (power-of-2 roundup).

pravindra commented on java/vector/src/main/java/org/apache/arrow/vector/BaseValueVector.java:
```diff
@@ -98,5 +105,67 @@ protected ArrowBuf releaseBuffer(ArrowBuf buffer) {
+    } else {
+      // Due to roundup to power-of-2 allocation, the bufferSize could be greater than the
+      // requested size. Utilize the allocated buffer fully.;
+      int actualCount = (int)((bufferSize * 8.0) / (8 * typeWidth + 1));
```
will fix

pravindra commented:
> > You're commit message doesn't match your changes. Why are you making the other changes?
> 
> My initial change reduced the heap usage only for fixed-width vectors. This change extends the same fix for variable-width vectors, and the restructuring for code re-use. I'll open a separate lira for the variable-width vectors.

I've opened ARROW-4147 for the rest of the change. Will use this to fix the race issue only.

pravindra commented:
@siddharthteotia @jacques-n - this PR now only has the fix for the race condition. Can you please merge this change ?
